### ISSUE TITLE
StreamPeerSSL not returning any data

### ISSUE THREAD

**Issue by iggyvolz** [godotengine/godot#14970]
**Godot version:**
3.0 Beta 2 Snapshot (https://godotengine.org/article/dev-snapshot-godot-3-0-beta-2) - Mono version

**OS/device including version:**
Windows 10 Education Build 16299, 64-bit

**Issue description:**
StreamPeerSSL is never producing any output, even though an identical TCP connection returns output.

**Steps to reproduce:**
Simple script: https://gist.github.com/iggyvolz/53088998c4b9cbdb525252aa5a025618

**Minimal reproduction project:**
[BrokenSSL.zip](https://github.com/godotengine/godot/files/1583835/BrokenSSL.zip)

- [X] I searched the existing [GitHub issues](https://github.com/godotengine/godot/issues?utf8=%E2%9C%93&q=is%3Aissue+) for potential duplicates.

**Comment by akien-mga** [comment]
CC @Faless

**Comment by Faless** [comment]
I can't run the mono thing, but I'm pretty sure the bug is in the return code of `ConnectToStream` which should fail as the TCP stream is surely not connected at that time.
Try and wait for the TCP stream to be connected before calling `ConnectToStream`.
I never touched the SSL part, but I guess https://github.com/godotengine/godot/blob/master/modules/openssl/stream_peer_openssl.cpp#L311 needs few fixes on the way it sets the status.

**Comment by Faless** [comment]
Okay, some updates here, there are 3 problems:

1. StreamPeerSSL does not always properly set the status when connecting to stream (easy to fix, already have it locally)

2. StreamPeerSSL only does blocking I/O, which locks the game if you do something like `get_partial_data(1)` and there are no bytes available (should only lock when doing `get_data(1)` and no bytes are available). This is harder to fix, I don't have a solution for now, I'm looking into it.

3. `get_available_bytes()` is not reliable, it always returns 0 until you do a read. At that point it will return the leftover bytes but not any new byte coming after the read. Properly need to fix number 2 before working on this

**Comment by reduz** [comment]
@Faless 

Any status on this?

Fix on 1. should be good
Fix on 2. I tried but I failed to find a way to do this, so it always locks for now.
Fix on 3 is as you say, but I just cant figure out how to fix 2, it may not be possible unless moving to a thread or something like that.

**Comment by Faless** [comment]
@reduz 
1. Done locally
2. I'll try to work on it today and tomorrow, see if I can find a way to make it work.
3. I guess it's dependent on 2

Will keep you posted.

**Comment by Faless** [comment]
@reduz 
I think the only way for `get_available_bytes()` to work is to add a `poll` method to `StreamPeerSSL` that pre-fetches the data in a `RingBuffer`.
The `poll` method will be called automatically for `get_data` and `get_partial_data` but MUST be called manually before `get_available_bytes` (as it's `cont`).

What do you think? I have it working locally and does not seem to break other stuff (eg. `HTTPClient` and AssetLib), I'll clean it up and PR it so we can discuss it better

**Comment by reduz** [comment]
Will prefer this is merged immediately after 3.0 is out, as it will need testing and it's not critical.

**Comment by beniwtv** [comment]
I got just hit by this bug I think, it seems StreamPeerSSL never connects. I do get an error in the debugger when calling `connect_to_stream()` though:

```
"Some I/O error occoured. The OpenSSL error queue may contain more information on the error"
modules/openssl/stream_peer_openssl.cpp:434
```

I made sure to wait till StreamPeerTCP was connected via `is_connected_to_host()`, made no difference.

**Comment by Faless** [comment]
@beniwtv `StreamPeerTCP.is_connected_to_host()` returns true even when the peer is still connecting.

Non blocking I/O and better status reporting in StreamPeerSSL are addressed by #16519.
`StreamPeerSSL` behaves correctly when connecting to a valid (`STATUS_CONNECTED`) TCP stream.

Closing this.
